contrib/format-excel/src/main/java/org/apache/drill/exec/store/excel/ExcelBatchReader.java

```diff
@@ -0,0 +1,526 @@
+      columnCount = sheet.getRow(sheet.getFirstRowNum()).getPhysicalNumberOfCells();
+    } else {
+      // Case for when the user defines the headerRow as -1 IE:  When there isn't a headerRow.
+      columnCount = sheet.getRow(0).getPhysicalNumberOfCells();
+    }
+    return columnCount;
+  }
+
+  @Override
+  public boolean next() {
+    recordCount = 0;
+    while (!rowWriter.isFull()) {
+      if (!nextLine(rowWriter)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private boolean nextLine(RowSetLoader rowWriter) {
+    if( sheet.getFirstRowNum() == 0 && sheet.getLastRowNum() == 0) {
+      // Case for empty sheet
+      return false;
+    } else if (!rowIterator.hasNext()) {
+      return false;
+    } else if (recordCount >= readerConfig.lastRow) {
+      return false;
+    }
+
+    int lastRow = readerConfig.lastRow;
+    while (recordCount < lastRow && rowIterator.hasNext()) {
+      lineCount++;
+
+      Row row = rowIterator.next();
+
+      // If the user specified that there are no headers, get the column count
+      if (readerConfig.headerRow == -1 && recordCount == 0) {
+        this.totalColumnCount = row.getLastCellNum();
+      }
+
+      int colPosition = 0;
+      if (readerConfig.firstColumn != 0) {
+        colPosition = readerConfig.firstColumn - 1;
+      }
+
+      int finalColumn = totalColumnCount;
+      if (readerConfig.lastColumn != 0) {
+        finalColumn = readerConfig.lastColumn - 1;
+      }
+      rowWriter.start();
+      for (int colWriterIndex = 0; colPosition < finalColumn; colPosition++) {
+        Cell cell = row.getCell(colPosition);
+
+        CellValue cellValue = evaluator.evaluate(cell);
+
+        populateColumnArray(cell, cellValue, colPosition);
+        cellWriterArray.get(colWriterIndex).load(cell);
+
+        colWriterIndex++;
+      }
+
+      if (firstLine) {
+        firstLine = false;
+      }
+      rowWriter.save();
+      recordCount++;
+    }
+    return true;
+  }
+
+  /**
+   * Function to populate the column array
+   * @param cell
+   * @param cellValue
+   * @param colPosition
+   */
+  private void populateColumnArray(Cell cell, CellValue cellValue, int colPosition) {
+    if (!firstLine) {
+      return;
+    }
+
+    if (cellValue == null) {
+      addColumnToArray(rowWriter, excelFieldNames.get(colPosition), MinorType.VARCHAR);
+    } else {
+      CellType cellType = cellValue.getCellTypeEnum();
+      if (cellType == CellType.STRING || readerConfig.readAllFieldsAsVarChar) {
+        addColumnToArray(rowWriter, excelFieldNames.get(colPosition), MinorType.VARCHAR);
+      } else if (cellType == CellType.NUMERIC && DateUtil.isCellDateFormatted(cell)) {
+        addColumnToArray(rowWriter, excelFieldNames.get(colPosition), MinorType.TIMESTAMP);
+      } else if (cellType == CellType.NUMERIC) {
+        addColumnToArray(rowWriter, excelFieldNames.get(colPosition), MinorType.FLOAT8);
+      } else {
+        logger.warn("Unknown data type.");
```

Comment:
Include cell type into want message...
If type is not supported which type it would have?

Reply:
There are really only two data types, `NUMERIC` and `STRING`.   I don't think this case will ever happen.